### ISSUE TITLE
co19/LanguageFeatures/Constructor-tear-offs/type_literal_A01_t01 issue

### ISSUE THREAD

**Issue by eernstg** [dart-lang/co19#1167]
Said test expects a newly created list, `List<int>.empty()`, to have run-time type `List<int>`. There is no reason to expect this, and it is actually easy to see that it could never be true for a "non-magic" (user-written) class because `class List` is abstract.

The language specification does not mention many details of the class `List`, and in particular it does not specify anything about the constructor which is used to obtain the new list. But by the general language rules we know that the constructor must be a factory, and the instance creation must yield an object whose dynamic type is some _subtype_ of `List<int>`. So the test could check that the list `is List<int>`.

The test cases using `Expect.isFalse` are somewhat misleading, so they may need to be adjusted as well.

**Comment by alexmarkov** [comment]
Note that `Object.runtimeType` documentation does not mention that it cannot return an abstract type:

```dart
  /// A representation of the runtime type of the object.
  external Type get runtimeType;
```

An implementation may hide its internal classes which implement built-in lists, so it may still return `List` type as a `.runtimeType` for built-in list objects. This is similar to other abstract types returned from `.runtimeType` such as `double`, `int` and `String`.

So the new expectation in the test still doesn't look right:

https://github.com/dart-lang/co19/blob/eccf5c88942776498ca8cc298c35a667263b0ad9/LanguageFeatures/Constructor-tear-offs/type_literal_A01_t01.dart#L28-L29

**Comment by eernstg** [comment]
Right, there is nothing stopping any given class from overriding `runtimeType` and returning any `Type` whatsoever, and we just need to consider the new syntax `List<int>` to see that we can obtain a `Type` for an abstract type.

With `dart2js` I get 'JSArray<dynamic>' from `print([].runtimeType)`, and the vm prints `List<dynamic>`. However, we get 'false' from the following:

```dart
bool sameType<X>(Type y) => X == y;

void main() {
  print([].runtimeType); // 'List<dynamic>'.
  print(sameType<List<dynamic>>([].runtimeType)); // 'false'!
}
```

I think the VM returns a specialized `Type` instance that prints as `'List<T>'` for whatever `T` the given list has, but it is actually just the `toString` of that `Type` that makes it similar to `List<T>`, the given `Type` does in fact represent some internal implementation class.

So it looks like we _do_ get a representation of the actual run-time type of built-in classes like `List`, even though it superficially looks like we get the abstract supertype `List` in some cases.

However, we can't even rely on that. Numerical types have the following:

```
lib/_internal/js_runtime/lib/js_number.dart:480:  Type get runtimeType => num;
lib/_internal/js_runtime/lib/js_number.dart:728:  Type get runtimeType => int;
lib/_internal/js_runtime/lib/js_number.dart:736:  Type get runtimeType => double;
```

So we'll get the run-time type `int` from an integer even though it could have any one of several internal representations.

In other words, we have to be really careful about assuming that `runtimeType` returns anything specific (which is even more true for system classes that may use various "magic" to make things look in a particular way), and we don't have support from the specification to _require_ anything specific.

That makes it a bit tricky to test _anything_ about the value of the expression `List<int>` (which, I think, is the point of this test overall), but we can still refer to the old-fashioned way to obtain such `Type` objects:

```dart
Type typeOf<X>() => X;

main() {
  Type intList = List<int>;
  Expect.isTrue(typeOf<List<int>>() == intList);
  Expect.isFalse(typeOf<List<double>>() == intList);
  Expect.isFalse(typeOf<List<num>>() == intList);
}
```
